omc3/knob_extractor.py

```diff
@@ -0,0 +1,459 @@
+r"""
+Knob Extractor
+--------------
+
+Will extract knobs and give information about the current beam process (no it doesn't?).
+Fetches data from nxcals through pytimber using the StateTracker fields.
+
+**Arguments:**
+
+*--Optional--*
+
+- **knob_definitions** *(PathOrStrOrDataFrame)*:
+
+    User defined path to the knob-definitions, or (via python) a dataframe
+    containing the knob definitions with the columns 'madx', 'lsa' and
+    'scaling'.
+
+
+- **knobs** *(str)*:
+
+    A list of knob names or categories to extract. Available categories
+    are: sep, xing, chroma, ip_offset, disp, mo.
+
+    default: ``['sep', 'xing', 'chroma', 'ip_offset', 'disp', 'mo']``
+
+
+- **output** *(PathOrStr)*:
+
+    Specify user-defined output path. This should probably be
+    `model_dir/knobs.madx`
+
+    default: ``knobs.madx``
+
+
+- **state**:
+
+    Prints the state of the StateTracker. Does not extract anything else.
+
+    action: ``store_true``
+
+
+- **time** *(str)*:
+
+    At what time to extract the knobs. Accepts ISO-format (YYYY-MM-
+    DDThh:mm:ss), timestamp or 'now'. The default timezone for the ISO-
+    format is local time, but you can force e.g. UTC by adding +00:00.
+
+    default: ``now``
+
+
+- **timedelta** *(str)*:
+
+    Add this timedelta to the given time. The format of timedelta is
+    '((\d+)(\w))+' with the second token being one of s(seconds),
+    m(minutes), h(hours), d(days), w(weeks), M(months) e.g 7m = 7 minutes,
+    1d = 1day, 7m30s = 7 min 30 secs. A prefix '_' specifies a negative
+    timedelta. This allows for easily getting the setting e.g. 2h ago:
+    '_2h' while setting the `time` argument to 'now' (default).
+
+
+"""
+import argparse
+import math
+import re
+from dataclasses import dataclass
+from datetime import datetime
+from pathlib import Path
+from typing import Dict, Sequence, Union, Optional, List, Tuple
+
+import pandas as pd
+import tfs
+from dateutil.relativedelta import relativedelta
+
+from generic_parser import EntryPointParameters, entrypoint
+from omc3.utils.iotools import PathOrStrOrDataFrame, PathOrStr
+from omc3.utils.logging_tools import get_logger
+from omc3.utils.mock import cern_network_import
+
+LOGGER = get_logger(__name__)
+
+pytimber = cern_network_import("pytimber")
```

Comment:
If I understand correctly, this still tries to import `pytimber` directly. In that case, this should be moved down

Reply:
moved up?

Reply:
maybe we can add a lazy initialisation to the mocking ` pytimber.LoggingDB()` function

Reply:
I don't understand. This should be handled like a normal import, no?

Reply:
@fsoubelet, what Andreas meant to say is that pytimber starts pulling java libraries when imported (or when LoggingDB called?)
This means you'd have to wait quite a long time, only to check if your arguments make sense for example

Reply:
It's only when LoggingDB is called. I checked. So import of pytimber should not interrupt workflow. I think what you are describing was a problem in earlier versions and might still be on in pjLSA, but not sure. Moved it up now to underline that it's an import.